File: src/domain/categories.ts
import type { SearchCategory, SearchSubcategory } from "./search-types.js";
import { normalizeText, uniqueSorted } from "../shared/utils.js";

export const SEARCH_CATEGORIES = [
  "equipment",
  "feat",
  "creature",
  "hazard",
  "affliction",
  "rule",
  "spell",
  "characterCreation",
  "lore",
] as const satisfies readonly SearchCategory[];

export const CATEGORY_SUBCATEGORY_MAP: Record<SearchCategory, SearchSubcategory[]> = {
  equipment: ["consumable", "gear", "weapon", "armor", "shield", "ammo", "backpack", "treasure", "kit", "vehicle"],
  feat: ["class", "ancestry", "skill", "general", "archetype", "boonCurse"],
  creature: ["character", "familiar"],
  hazard: ["haunt", "trap"],
  affliction: ["curse", "disease", "poison"],
  rule: ["action", "condition", "effect", "campaignFeature"],
  spell: [],
  characterCreation: ["ancestry", "heritage", "background", "class"],
  lore: ["deity", "journal"],
};

export const SEARCH_SUBCATEGORIES = uniqueSorted(Object.values(CATEGORY_SUBCATEGORY_MAP).flat());

const CATEGORY_INPUT_ALIASES = new Map<string, SearchCategory>([
  ...SEARCH_CATEGORIES.map((category) => [normalizeText(category), category] as const),
  ["feats", "feat"],
  ["creatures", "creature"],
  ["hazards", "hazard"],
  ["afflictions", "affliction"],
  ["rules", "rule"],
  ["spells", "spell"],
  ["character creation", "characterCreation"],
]);

const SUBCATEGORY_INPUT_ALIASES = new Map<string, SearchSubcategory>([
  ...SEARCH_SUBCATEGORIES.map((subcategory) => [normalizeText(subcategory), subcategory] as const),
  ["actions", "action"],
  ["conditions", "condition"],
  ["effects", "effect"],
  ["campaign", "campaignFeature"],
  ["campaigns", "campaignFeature"],
  ["campaign feature", "campaignFeature"],
  ["campaign features", "campaignFeature"],
  ["consumables", "consumable"],
  ["weapons", "weapon"],
  ["shields", "shield"],
  ["backpacks", "backpack"],
  ["treasures", "treasure"],
  ["kits", "kit"],
  ["vehicles", "vehicle"],
  ["archetypes", "archetype"],
  ["haunts", "haunt"],
  ["traps", "trap"],
  ["curses", "curse"],
  ["diseases", "disease"],
  ["poisons", "poison"],
  ["classes", "class"],
  ["ancestries", "ancestry"],
  ["heritages", "heritage"],
  ["backgrounds", "background"],
  ["deities", "deity"],
  ["journals", "journal"],
]);

const SUBCATEGORY_TO_CATEGORIES = new Map<string, SearchCategory[]>(
  SEARCH_SUBCATEGORIES.map((subcategory) => {
    const categories = SEARCH_CATEGORIES.filter((category) => CATEGORY_SUBCATEGORY_MAP[category].includes(subcategory));
    return [normalizeText(subcategory), categories] as const;
  }),
);

export const VALID_SEARCH_CATEGORY_LIST = SEARCH_CATEGORIES.join(", ");
export const VALID_SEARCH_SUBCATEGORY_LIST = SEARCH_SUBCATEGORIES.join(", ");

export function normalizeSearchCategory(value: string | null | undefined): SearchCategory | null {
  const normalized = normalizeText(value ?? "");
  return CATEGORY_INPUT_ALIASES.get(normalized) ?? null;
}

export function normalizeSearchSubcategory(value: string | null | undefined): SearchSubcategory | null {
  const normalized = normalizeText(value ?? "");
  return SUBCATEGORY_INPUT_ALIASES.get(normalized) ?? null;
}

export function getSearchCategoryErrorMessage(value: string): string {
  return `Unknown top-level category "${value}". Valid categories: ${VALID_SEARCH_CATEGORY_LIST}.`;
}

export function getSearchSubcategoryErrorMessage(value: string): string {
  return `Unknown subcategory "${value}". Valid subcategories: ${VALID_SEARCH_SUBCATEGORY_LIST}.`;
}

export function categorySupportsSubcategory(category: SearchCategory, subcategory: SearchSubcategory): boolean {
  return CATEGORY_SUBCATEGORY_MAP[category].includes(subcategory);
}

export function getCategoryForSubcategory(subcategory: string | null | undefined): SearchCategory | null {
  const canonicalSubcategory = normalizeSearchSubcategory(subcategory);
  if (!canonicalSubcategory) {
    return null;
  }

  const categories = SUBCATEGORY_TO_CATEGORIES.get(normalizeText(canonicalSubcategory)) ?? [];
  return categories.length === 1 ? categories[0]! : null;
}

export function getCategoryKeywordAnchors(): Record<SearchCategory, string[]> {
  return {
    creature: ["creature", "creatures", "monster", "monsters", "beast", "undead", "swarm", "dragon", "fiend", "enemy"],
    hazard: ["hazard", "hazards", "haunt", "haunted", "trap", "traps", "snare", "snares"],
    spell: [
      "spell",
      "spells",
      "cantrip",
      "cantrips",
      "ritual",
      "rituals",
      "focus",
      "arcane",
      "divine",
      "occult",
      "primal",
      "magic",
      "magical",
    ],
    equipment: [
      "gear",
      "equipment",
      "item",
      "items",
      "weapon",
      "weapons",
      "armor",
      "shield",
      "consumable",
      "consumables",
      "potion",
      "potions",
      "elixir",
      "elixirs",
      "bomb",
      "bombs",
      "gadget",
      "gadgets",
      "ammo",
      "treasure",
    ],
    feat: [
      "feat",
      "feats",
      "dedication",
      "archetype",
      "general feat",
      "skill feat",
      "class feat",
      "ancestry feat",
      "boon",
      "boons",
      "curse feat",
    ],
    affliction: [
      "affliction",
      "afflictions",
      "disease",
      "diseases",
      "poison",
      "poisons",
      "venom",
      "venoms",
      "toxin",
      "toxins",
      "plague",
      "plagues",
      "curse",
      "curses",
    ],
    rule: [
      "rule",
      "rules",
      "action",
      "actions",
      "condition",
      "conditions",
      "effect",
      "effects",
      "trait",
      "traits",
      "kingdom",
      "warfare",
      "army",
      "subsystem",
    ],
    characterCreation: ["character creation", "ancestry", "heritage", "background", "backgrounds", "class", "classes"],
    lore: ["lore", "setting", "deity", "deities", "god", "gods", "pantheon", "faith", "religion", "journal"],
  };
}

export function getSubcategoryKeywordAnchors(): Array<{
  subcategory: string;
  category: SearchCategory;
  keywords: string[];
  weight?: number;
}> {
  return [
    { category: "hazard", subcategory: "haunt", keywords: ["haunt", "haunted", "manifestation", "poltergeist"] },
    { category: "hazard", subcategory: "trap", keywords: ["trap", "traps", "snare", "snares", "tripwire", "pitfall"] },
    {
      category: "equipment",
      subcategory: "consumable",
      keywords: [
        "consumable",
        "consumables",
        "potion",
        "potions",
        "elixir",
        "elixirs",
        "bomb",
        "bombs",
        "fulu",
        "mutagen",
        "mutagens",
      ],
    },
    { category: "equipment", subcategory: "weapon", keywords: ["weapon", "weapons"] },
    { category: "equipment", subcategory: "armor", keywords: ["armor"] },
    { category: "equipment", subcategory: "shield", keywords: ["shield", "shields"] },
    {
      category: "equipment",
      subcategory: "ammo",
      keywords: ["ammo", "ammunition", "arrow", "arrows", "bolt", "bolts"],
    },
    { category: "equipment", subcategory: "treasure", keywords: ["treasure", "loot", "hoard"] },
    { category: "feat", subcategory: "class", keywords: ["class feat", "class feats"], weight: 2 },
    { category: "feat", subcategory: "ancestry", keywords: ["ancestry feat", "ancestry feats"], weight: 2 },
    { category: "feat", subcategory: "skill", keywords: ["skill feat", "skill feats"], weight: 2 },
    { category: "feat", subcategory: "general", keywords: ["general feat", "general feats"], weight: 2 },
    {
      category: "feat",
      subcategory: "archetype",
      keywords: ["archetype", "dedication", "archetype feat", "archetype feats"],
    },
    { category: "feat", subcategory: "boonCurse", keywords: ["boon", "boons", "curse feat", "curse feats"] },
    { category: "rule", subcategory: "action", keywords: ["action", "actions"] },
    { category: "rule", subcategory: "condition", keywords: ["condition", "conditions"] },
    { category: "rule", subcategory: "effect", keywords: ["effect", "effects"] },
    {
      category: "rule",
      subcategory: "campaignFeature",
      keywords: [
        "kingdom",
        "warfare",
        "army",
        "settlement",
        "hex",
        "campaign",
        "campaign feature",
        "campaign features",
      ],
      weight: 3,
    },
    { category: "characterCreation", subcategory: "ancestry", keywords: ["ancestry"] },
    { category: "characterCreation", subcategory: "heritage", keywords: ["heritage"] },
    { category: "characterCreation", subcategory: "background", keywords: ["background", "backgrounds"] },
    { category: "characterCreation", subcategory: "class", keywords: ["class", "classes"] },
    {
      category: "lore",
      subcategory: "deity",
      keywords: ["deity", "deities", "god", "gods", "pantheon", "faith", "religion"],
    },
    { category: "lore", subcategory: "journal", keywords: ["journal", "history", "setting", "lore"] },
    { category: "affliction", subcategory: "disease", keywords: ["disease", "diseases", "plague", "plagues"] },
    {
      category: "affliction",
      subcategory: "poison",
      keywords: ["poison", "poisons", "venom", "venoms", "toxin", "toxins"],
    },
    { category: "affliction", subcategory: "curse", keywords: ["curse", "curses"] },
  ];
}

File: rust/crates/atlas-domain/src/categories.rs
use std::fmt;
use std::str::FromStr;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Category {
    Equipment,
    Feat,
    Creature,
    Hazard,
    Affliction,
    Rule,
    Spell,
    CharacterCreation,
    Lore,
}

impl Category {
    pub const ALL: [Self; 9] = [
        Self::Equipment,
        Self::Feat,
        Self::Creature,
        Self::Hazard,
        Self::Affliction,
        Self::Rule,
        Self::Spell,
        Self::CharacterCreation,
        Self::Lore,
    ];

    pub fn as_str(self) -> &'static str {
        match self {
            Self::Equipment => "equipment",
            Self::Feat => "feat",
            Self::Creature => "creature",
            Self::Hazard => "hazard",
            Self::Affliction => "affliction",
            Self::Rule => "rule",
            Self::Spell => "spell",
            Self::CharacterCreation => "character_creation",
            Self::Lore => "lore",
        }
    }

    pub fn from_canonical(value: &str) -> Option<Self> {
        match value {
            "equipment" => Some(Self::Equipment),
            "feat" => Some(Self::Feat),
            "creature" => Some(Self::Creature),
            "hazard" => Some(Self::Hazard),
            "affliction" => Some(Self::Affliction),
            "rule" => Some(Self::Rule),
            "spell" => Some(Self::Spell),
            "character_creation" => Some(Self::CharacterCreation),
            "lore" => Some(Self::Lore),
            _ => None,
        }
    }

    pub fn from_input(value: &str) -> Option<Self> {
        match normalize_input(value).as_str() {
            "equipment" => Some(Self::Equipment),
            "feat" | "feats" => Some(Self::Feat),
            "creature" | "creatures" => Some(Self::Creature),
            "hazard" | "hazards" => Some(Self::Hazard),
            "affliction" | "afflictions" => Some(Self::Affliction),
            "rule" | "rules" => Some(Self::Rule),
            "spell" | "spells" => Some(Self::Spell),
            "charactercreation" | "character creation" => Some(Self::CharacterCreation),
            "lore" => Some(Self::Lore),
            _ => None,
        }
    }

    pub fn subcategories(self) -> &'static [Subcategory] {
        match self {
            Self::Equipment => &[
                Subcategory::Consumable,
                Subcategory::Gear,
                Subcategory::Weapon,
                Subcategory::Armor,
                Subcategory::Shield,
                Subcategory::Ammo,
                Subcategory::Backpack,
                Subcategory::Treasure,
                Subcategory::Kit,
                Subcategory::Vehicle,
            ],
            Self::Feat => &[
                Subcategory::Class,
                Subcategory::Ancestry,
                Subcategory::Skill,
                Subcategory::General,
                Subcategory::Archetype,
                Subcategory::BoonCurse,
            ],
            Self::Creature => &[Subcategory::Character, Subcategory::Familiar],
            Self::Hazard => &[Subcategory::Haunt, Subcategory::Trap],
            Self::Affliction => &[
                Subcategory::Curse,
                Subcategory::Disease,
                Subcategory::Poison,
            ],
            Self::Rule => &[
                Subcategory::Action,
                Subcategory::Condition,
                Subcategory::Effect,
                Subcategory::CampaignFeature,
            ],
            Self::Spell => &[],
            Self::CharacterCreation => &[
                Subcategory::Ancestry,
                Subcategory::Heritage,
                Subcategory::Background,
                Subcategory::Class,
            ],
            Self::Lore => &[Subcategory::Deity, Subcategory::Journal],
        }
    }

    pub fn supports_subcategory(self, subcategory: Subcategory) -> bool {
        self.subcategories().contains(&subcategory)
    }
}

impl fmt::Display for Category {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.as_str())
    }
}

impl FromStr for Category {
    type Err = CategoryParseError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        Self::from_canonical(value).ok_or_else(|| CategoryParseError {
            value: value.to_string(),
        })
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CategoryParseError {
    value: String,
}

impl fmt::Display for CategoryParseError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "unknown category `{}`", self.value)
    }
}

impl std::error::Error for CategoryParseError {}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Subcategory {
    Consumable,
    Gear,
    Weapon,
    Armor,
    Shield,
    Ammo,
    Backpack,
    Treasure,
    Kit,
    Vehicle,
    Class,
    Ancestry,
    Skill,
    General,
    Archetype,
    BoonCurse,
    Character,
    Familiar,
    Haunt,
    Trap,
    Curse,
    Disease,
    Poison,
    Action,
    Condition,
    Effect,
    CampaignFeature,
    Heritage,
    Background,
    Deity,
    Journal,
}

impl Subcategory {
    pub const ALL: [Self; 31] = [
        Self::Action,
        Self::Ammo,
        Self::Ancestry,
        Self::Archetype,
        Self::Armor,
        Self::Backpack,
        Self::Background,
        Self::BoonCurse,
        Self::CampaignFeature,
        Self::Character,
        Self::Class,
        Self::Condition,
        Self::Consumable,
        Self::Curse,
        Self::Deity,
        Self::Disease,
        Self::Effect,
        Self::Familiar,
        Self::Gear,
        Self::General,
        Self::Haunt,
        Self::Heritage,
        Self::Journal,
        Self::Kit,
        Self::Poison,
        Self::Shield,
        Self::Skill,
        Self::Trap,
        Self::Treasure,
        Self::Vehicle,
        Self::Weapon,
    ];

    pub fn as_str(self) -> &'static str {
        match self {
            Self::Consumable => "consumable",
            Self::Gear => "gear",
            Self::Weapon => "weapon",
            Self::Armor => "armor",
            Self::Shield => "shield",
            Self::Ammo => "ammo",
            Self::Backpack => "backpack",
            Self::Treasure => "treasure",
            Self::Kit => "kit",
            Self::Vehicle => "vehicle",
            Self::Class => "class",
            Self::Ancestry => "ancestry",
            Self::Skill => "skill",
            Self::General => "general",
            Self::Archetype => "archetype",
            Self::BoonCurse => "boon_curse",
            Self::Character => "character",
            Self::Familiar => "familiar",
            Self::Haunt => "haunt",
            Self::Trap => "trap",
            Self::Curse => "curse",
            Self::Disease => "disease",
            Self::Poison => "poison",
            Self::Action => "action",
            Self::Condition => "condition",
            Self::Effect => "effect",
            Self::CampaignFeature => "campaign_feature",
            Self::Heritage => "heritage",
            Self::Background => "background",
            Self::Deity => "deity",
            Self::Journal => "journal",
        }
    }

    pub fn from_canonical(value: &str) -> Option<Self> {
        match value {
            "action" => Some(Self::Action),
            "ammo" => Some(Self::Ammo),
            "ancestry" => Some(Self::Ancestry),
            "archetype" => Some(Self::Archetype),
            "armor" => Some(Self::Armor),
            "backpack" => Some(Self::Backpack),
            "background" => Some(Self::Background),
            "boon_curse" => Some(Self::BoonCurse),
            "campaign_feature" => Some(Self::CampaignFeature),
            "character" => Some(Self::Character),
            "class" => Some(Self::Class),
            "condition" => Some(Self::Condition),
            "consumable" => Some(Self::Consumable),
            "curse" => Some(Self::Curse),
            "deity" => Some(Self::Deity),
            "disease" => Some(Self::Disease),
            "effect" => Some(Self::Effect),
            "familiar" => Some(Self::Familiar),
            "gear" => Some(Self::Gear),
            "general" => Some(Self::General),
            "haunt" => Some(Self::Haunt),
            "heritage" => Some(Self::Heritage),
            "journal" => Some(Self::Journal),
            "kit" => Some(Self::Kit),
            "poison" => Some(Self::Poison),
            "shield" => Some(Self::Shield),
            "skill" => Some(Self::Skill),
            "trap" => Some(Self::Trap),
            "treasure" => Some(Self::Treasure),
            "vehicle" => Some(Self::Vehicle),
            "weapon" => Some(Self::Weapon),
            _ => None,
        }
    }

    pub fn from_input(value: &str) -> Option<Self> {
        match normalize_input(value).as_str() {
            "action" | "actions" => Some(Self::Action),
            "ammo" => Some(Self::Ammo),
            "ancestry" | "ancestries" => Some(Self::Ancestry),
            "archetype" | "archetypes" => Some(Self::Archetype),
            "armor" => Some(Self::Armor),
            "backpack" | "backpacks" => Some(Self::Backpack),
            "background" | "backgrounds" => Some(Self::Background),
            "booncurse" => Some(Self::BoonCurse),
            "campaign" | "campaigns" | "campaign feature" | "campaign features"
            | "campaignfeature" => Some(Self::CampaignFeature),
            "character" => Some(Self::Character),
            "class" | "classes" => Some(Self::Class),
            "condition" | "conditions" => Some(Self::Condition),
            "consumable" | "consumables" => Some(Self::Consumable),
            "curse" | "curses" => Some(Self::Curse),
            "deity" | "deities" => Some(Self::Deity),
            "disease" | "diseases" => Some(Self::Disease),
            "effect" | "effects" => Some(Self::Effect),
            "familiar" => Some(Self::Familiar),
            "gear" => Some(Self::Gear),
            "general" => Some(Self::General),
            "haunt" | "haunts" => Some(Self::Haunt),
            "heritage" | "heritages" => Some(Self::Heritage),
            "journal" | "journals" => Some(Self::Journal),
            "kit" | "kits" => Some(Self::Kit),
            "poison" | "poisons" => Some(Self::Poison),
            "shield" | "shields" => Some(Self::Shield),
            "skill" => Some(Self::Skill),
            "trap" | "traps" => Some(Self::Trap),
            "treasure" | "treasures" => Some(Self::Treasure),
            "vehicle" | "vehicles" => Some(Self::Vehicle),
            "weapon" | "weapons" => Some(Self::Weapon),
            _ => None,
        }
    }
}

impl fmt::Display for Subcategory {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.as_str())
    }
}

impl FromStr for Subcategory {
    type Err = SubcategoryParseError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        Self::from_canonical(value).ok_or_else(|| SubcategoryParseError {
            value: value.to_string(),
        })
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SubcategoryParseError {
    value: String,
}

impl fmt::Display for SubcategoryParseError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "unknown subcategory `{}`", self.value)
    }
}

impl std::error::Error for SubcategoryParseError {}

fn normalize_input(value: &str) -> String {
    value.trim().to_lowercase()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn category_serializes_to_rust_canonical_values() {
        assert_eq!(
            serde_json::to_string(&Category::CharacterCreation).expect("category serializes"),
            "\"character_creation\""
        );
    }

    #[test]
    fn category_accepts_current_ts_aliases() {
        assert_eq!(
            Category::from_input("character creation"),
            Some(Category::CharacterCreation)
        );
        assert_eq!(Category::from_input("feats"), Some(Category::Feat));
        assert_eq!(Category::from_input("rules"), Some(Category::Rule));
    }

    #[test]
    fn subcategory_serializes_to_rust_canonical_values() {
        assert_eq!(
            serde_json::to_string(&Subcategory::BoonCurse).expect("subcategory serializes"),
            "\"boon_curse\""
        );
        assert_eq!(
            serde_json::to_string(&Subcategory::CampaignFeature).expect("subcategory serializes"),
            "\"campaign_feature\""
        );
    }

    #[test]
    fn subcategory_accepts_current_ts_aliases() {
        assert_eq!(
            Subcategory::from_input("actions"),
            Some(Subcategory::Action)
        );
        assert_eq!(
            Subcategory::from_input("campaign features"),
            Some(Subcategory::CampaignFeature)
        );
        assert_eq!(
            Subcategory::from_input("ancestries"),
            Some(Subcategory::Ancestry)
        );
    }

    #[test]
    fn category_reports_supported_subcategories() {
        assert!(Category::Equipment.supports_subcategory(Subcategory::Weapon));
        assert!(Category::CharacterCreation.supports_subcategory(Subcategory::Class));
        assert!(!Category::Spell.supports_subcategory(Subcategory::Action));
    }

    #[test]
    fn serde_accepts_only_rust_canonical_values() {
        let category: Category =
            serde_json::from_str("\"character_creation\"").expect("canonical value deserializes");
        assert_eq!(category, Category::CharacterCreation);
        assert_eq!(
            serde_json::to_string(&category).expect("category serializes"),
            "\"character_creation\""
        );
        assert!(serde_json::from_str::<Category>("\"characterCreation\"").is_err());
        assert!(serde_json::from_str::<Category>("\"feats\"").is_err());

        let subcategory: Subcategory =
            serde_json::from_str("\"campaign_feature\"").expect("canonical value deserializes");
        assert_eq!(subcategory, Subcategory::CampaignFeature);
        assert_eq!(
            serde_json::to_string(&subcategory).expect("subcategory serializes"),
            "\"campaign_feature\""
        );
        assert!(serde_json::from_str::<Subcategory>("\"campaignFeature\"").is_err());
        assert!(serde_json::from_str::<Subcategory>("\"campaign\"").is_err());
    }
}
